### PR TITLE
WIP: implement caching decorator

### PR DESCRIPTION
**Pull request by erdii**
this PR implements a cache decorator as requested in https://issues.redhat.com/browse/APPSRE-2866

### CODE REVIEW

**Comment by apahim** [comment]
Man, there's soo much to love in this PR :)
Reviewing...

File: sretoolbox/utils/cache.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
+"""
+Caching helper decorator
+"""
+
+import functools
+import weakref
+
+_CACHES = weakref.WeakKeyDictionary()
```
Comment (apahim) [inline sretoolbox/utils/cache.py:8]: I can't think on a case when the child dictionay will be garbage collected. Once you decora a function, the reference to that dict will last the entyre session. What am I missing?

File: sretoolbox/utils/cache.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+
+def static(key):
```
Comment (apahim) [inline sretoolbox/utils/cache.py:11]: I believe users should not have that much control over the cache... I can see people using the same keys in completly unrelated functions.
I know, you want a link between getters and setters, but asking users to do it is not the way to go, IMO.

File: sretoolbox/utils/cache.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    """
+    Decorates a class method to look up and cache the execution result.
+
+    Cache access is scoped by 2 things: class instance and `key`
+
+    :param key: static cache key
+    :type key: string
```
Comment (apahim) [inline sretoolbox/utils/cache.py:18]: docstrings <3

File: sretoolbox/utils/cache.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+            cache = _lazy_init(self)
+            value = None
+            try:
+                value = cache[key]
+            except KeyError:
+                value = func(self, *args, **kwargs)
+                cache[key] = value
+            return value
+        return with_cache_get
+    return decorate
```
Comment (apahim) [inline sretoolbox/utils/cache.py:32]: awesome

File: sretoolbox/utils/cache.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+
+
+def computed(key_func):
```
Comment (apahim) [inline sretoolbox/utils/cache.py:35]: the computed variant is giving users even more autonomy over the keys they use... I'd go in the oposite direction, but if you can convince me we need users providing the keys, then this function might come handy.

File: tests/test_cache.py [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+        """
+        mutate something - needing a computed cache invalidation afterwards
+        this returns a new map on every call - so that we can check for reference equality
+        """
+        return {
+            "collection_id": collection_id,
+            "item_id": item_id
+        }
+
+class TestCache:
```
Comment (apahim) [inline tests/test_cache.py:72]: Man, I can't express my appreciation enough. Thanks for the tests!

**Review by apahim — CHANGES_REQUESTED** [review]
@erdii, this PR is beatiful. Well documented, well tested and well implemented.

I just wanted to challenge the interface here. The typical use case is:

```python
class MyApiCli:
    def get_foo(self, a, b):
        # requests.get(), oc.get(), ...
        return 'bar'
    
    def set_foo(self, a, b):
        # requests.post(), oc.apply(), ...
        pass
```

I'm wondering if we can have an implementation that allows for no key control? Example:

```python
from ... import cache

class MyApiCli:
    @cache
    def get_foo(self, a, b):
        # requests.get(), oc.get(), ...
        return 'bar'

    @cache    
    def set_foo(self, a, b):
        # requests.post(), oc.apply(), ...
        pass
```
As long as the function signature is the same (apart from the `get_`/`set_` in the function name), both are using the same cache.

And because it's not always possible to keep get and set with the same signature, a direct interface for cache reset might come handy. Example:

```python
from ... import cache
from ... import reset_cache


class MyApiCli:

    @cache
    def read(self, secret, version):
        return 'bar'

    def write(self, secret):
        reset_cache(read)
        # or reset_cache(read, args, kwargs)
        ...
```

That would allow for a simple standard interface for caching with an explicit cache reset mechanism for when things can't be standard. On the other hand, it would require clients to adapt to the interface (which I like better than giving them control over the keys), so I'm not completly convinced.

In any case, the key should be unique per function+arguments. `get('foo')` can't return a result cached by `get('bar')`.

Wdyt?

File: sretoolbox/utils/cache.py (hunk 2 above)
Comment (apahim) [reply sretoolbox/utils/cache.py:11]: Anyway, key should be unique per function+arguments.

File: sretoolbox/utils/cache.py (hunk 1 above)
Comment (erdii) [reply sretoolbox/utils/cache.py:8]: The class instance using @cache.* decorators on its methods is used as a key to this WeakKeyDictionary and the value is its own scoped cache dict.

The cache dict will get garbage collected when the class instance is gone.

Comment (apahim) [reply sretoolbox/utils/cache.py:8]: Right! That makes sense and I like it :)
But then we are broken when we decorate a function (not a method in a class).

File: sretoolbox/utils/cache.py (hunk 2 above)
Comment (erdii) [reply sretoolbox/utils/cache.py:11]: I thought of implementing something like
- defaulting the key to the wrapped function's name
- defaulting the key to a derivate of the wrapped function's name like this:
```python
# `func` is the function we wrap/decorate
def _default_static_key(key, func):
    if key is not None:
        return key
    return re.sub(r'[get|set]_', '', func.__name__)
```

but IMHO we should provide a reasonable escape hatch to override this.
Example: 1 of the 2 links provided in the issue uses the function name `_read_all_v2`

---

> key should be unique per function+arguments

I tried to come up with a way for this but I was not able to create a  function `compute_cache_key(func, *args, **kwargs)` that is valid for all ways to call the decorated function.

given

```python
def foo(bar, baz):
  pass
```

i can call it in these ways

```python
foo("hello", "world") # *args = ("hello", "world"), **kwargs = {}
foo(bar="hello", baz="world") # *args = (), **kwargs = {"bar": "hello", "baz": "world"}
foo("hello", baz="world") # *args = ("hello"), **kwargs = {"baz": "world"})
```

which, when 'just' using a tuple of `(func.__name__, args, kwargs)`.

Do you have any thoughts on this?

File: sretoolbox/utils/cache.py (hunk 3 above)
Comment (erdii) [reply sretoolbox/utils/cache.py:18]: I'll add the param docs to all other functions as well

File: sretoolbox/utils/cache.py (hunk 2 above)
Comment (apahim) [reply sretoolbox/utils/cache.py:11]: Why do you care? If `foo("hello", "world")` and `foo(bar="hello", baz="world")` are using different backing caches, it's fine.

That will be yet another reason to be consistent . We can even improve ours docs here, including some remarks on that:
https://gitlab.cee.redhat.com/service/dev-guidelines/-/blob/master/python-style-guide.md#L326-352

But having `foo("hello", "world")` and a subsequent `foo("howdy", "folks")` returning the same value cannot happen.

File: sretoolbox/utils/cache.py (hunk 1 above)
Comment (erdii) [reply sretoolbox/utils/cache.py:8]: I see... very very good catch! I'll try to come up with a concept that works without the need for a class instance.

**Comment by erdii** [comment]
I have explored the proposal with automatic get_x/set_x cache keys and I'd advise against it.
The code a user has to write is awesomly slim, but it also hides magic behind it and ties them into a specific way of using the lib.


For example: 2 unrelated functions could share the same cache keys with no way to use another key (except renaming the functions but sometimes they have to conform to a specific naming pattern)
Several reasons:
- We deliberately hide cache keys which makes it difficult to share the same key betwen get/set/delete functions
- We trade having to specify a cache key against having to conform to a specific naming pattern
- We could only do the set-operation via decorator and have an asymmetric api for unsetting/updating <- this feels unergonomic to the user

- relying on string keys derived from function names could lead to very confusing behaviour (to be fair: this also applies to manually supplied string keys as long as the cache is shared globally and not scoped by something like a class instance - which is the case right now but will go away when I change the code to allow decorating bare functions)
```
class SomewhereInAProject:
    @cache
    def get_item(id):
        return {"id": id}

class SomewhereElseInTheProject:
    @cache
    def get_item(id):
        return {"itemId": id}
```

----

I'd like to clarify some points:
- what problems are we exactly trying to solve here?
- decorate methods/functions/both?
- testability of decorated stuff
  - caches tied to object instances via a global WeakRefDict are still somewhat testable
  - having a single global cache object restricts this in my eyes
  - we could also go the route of having a cache-class and instances that need to be passed around (clean but forces code changes)

**Comment by apahim** [comment]
> I have explored the proposal with automatic get_x/set_x cache keys and I'd advise against it.

Makes sense. Thanks looking into that.

> For example: 2 unrelated functions could share the same cache keys with no way to use another key (except renaming the functions but sometimes they have to conform to a specific naming pattern)

Not if the function/method object is part of the key (instead of the function name).

> what problems are we exactly trying to solve here?

- A common interface to cache the return of a given method/function.
- The ability to clear the cache data so the next call to the function that returns data will return freshly acquired data.

> decorate methods/functions/both?

Yes, both. We want to cache returns of functions and methods. And we want to be able to clear the cache to have fresh data being returned.

**Comment by erdii** [comment]
@apahim and me had a quick chat and we decided to go ahead with this interface approach:

```python
class MyApiCli:

    @cache
    def read(self, secret, version):
        return 'bar'

    def write(self, secret):
        # reset_cache()  # to clean the entire cache
        
        # optional if reasonable to implement
        # self.read.wipe_cache()
        # reset_cache(self.read)  # to clean the cache for self.read with any arg
```

- decreased testability in this case is not that important as we will not build huge applications on top of this simple cache
- we need to be able to also decorate bare functions -> no scoping by class instance
- we trade ergonomics and a symmetric interface for key safety by using the wrapped function object as part of the key
- we implement the bare minimum requirements:
  1. being able to cache the result of a function
  2. being able to invalidate a cached result (implementing this as a full cache flush is acceptable)

**Comment by erdii** [comment]
i just found https://docs.python.org/3/library/functools.html#functools.lru_cache
(note that `functools.cache` would be a better fit, but was not available before python 3.9)

```
import functools
 
# https://docs.python.org/3/library/functools.html#functools.lru_cache
@functools.lru_cache()
def foo(a, b):
	return {
		"a": a,
		"b": b
	}
 
def test():
	first = foo("a", "b")
	second = foo("a", "b")
	assert first is second
	third = foo("c", "b")
	assert first is not third
	
	foo.cache_clear()
	fourth = foo("a", "b")
	assert first is not fourth
 
if __name__ == "__main__":
    test()
```

**Comment by erdii** [comment]
closing as we will just use `functools.lru_cache`

**Pull request closed without merging**